src/class_decorator_downlevel_transformer.ts

```diff
@@ -0,0 +1,119 @@
+}
+
+function extractMetadataFromSingleDecorator(decorator: ts.Decorator) {
+  const metadataProperties: ts.ObjectLiteralElementLike[] = [];
+  const expr = decorator.expression;
+  switch (expr.kind) {
+    case ts.SyntaxKind.Identifier:
+      // The decorator was a plain @Foo.
+      metadataProperties.push(ts.createPropertyAssignment('type', expr));
+      break;
+    case ts.SyntaxKind.CallExpression:
+      // The decorator was a call, like @Foo(bar).
+      const call = expr as ts.CallExpression;
+      metadataProperties.push(ts.createPropertyAssignment('type', call.expression));
+      if (call.arguments.length) {
+        const args: ts.Expression[] = [];
+        for (const arg of call.arguments) {
+          args.push(arg);
+        }
+        const argsArrayLiteral = ts.createArrayLiteral(args);
+        argsArrayLiteral.elements.hasTrailingComma = true;
+        metadataProperties.push(ts.createPropertyAssignment('args', argsArrayLiteral));
+      }
+      break;
+    default:
+      throw new Error(`Unimplemented kind: ${ts.SyntaxKind[expr.kind]}`);
+  }
+  return ts.createObjectLiteral(metadataProperties);
+}
+
+function createDecoratorClassProperty(decoratorList: ts.ObjectLiteralExpression[]) {
+  const modifier = ts.createToken(ts.SyntaxKind.StaticKeyword);
+  const type = createClassDecoratorType();
+  const initializer = ts.createArrayLiteral(decoratorList, true);
+  initializer.elements.hasTrailingComma = true;
+  return ts.createProperty(undefined, [modifier], 'decorators', undefined, type, initializer);
+}
+
+function isNameEqual(classMember: ts.ClassElement, name: string): boolean {
+  if (classMember.name === undefined) {
+    return false;
+  }
+  const id = classMember.name as ts.Identifier;
+  return id.text === name;
+}
+
+function insertBeforeDecoratorProperties(
+    classMembers: ts.NodeArray<ts.ClassElement>, decoratorMetadata: ts.PropertyDeclaration) {
+  const ctorParameterIndex = classMembers.findIndex(m => isNameEqual(m, 'ctorParameters'));
+  const propDecoratorsIndex = classMembers.findIndex(m => isNameEqual(m, 'propDecorators'));
+
+  if (ctorParameterIndex === -1 && propDecoratorsIndex === -1) {
+    return;
+  }
+
+  let insertionPoint;
+  if (ctorParameterIndex === -1) {
+    insertionPoint = propDecoratorsIndex;
+  } else if (propDecoratorsIndex === -1) {
+    insertionPoint = ctorParameterIndex;
+  } else {
+    insertionPoint = Math.min(ctorParameterIndex, propDecoratorsIndex);
+  }
+  classMembers.splice(insertionPoint, 0, decoratorMetadata);
+}
+
+export function transformer(typeChecker: ts.TypeChecker): (context: ts.TransformationContext) =>
+    ts.Transformer<ts.SourceFile> {
+  return (context: ts.TransformationContext) => {
+    const visitor: ts.Visitor = (node: ts.Node): ts.Node => {
+      switch (node.kind) {
+        case ts.SyntaxKind.ClassDeclaration:
+          const cd = ts.visitEachChild(node, visitor, context) as ts.ClassDeclaration;
```

Comment:
Why is `cd` mutable here? Wouldn't you normally need to create a mutable clone?

Reply:
Are you sure? Wouldn't that mean ts.visitEachChild will always copies the entire tree?

The docs say: `possibly returning a new Node of the same kind in its place` - that doesn't sounds as if you can rely on that behaviour, does it?

Reply:
You're right, we do need to call `ts.getMutableClone` - I talked to Tobias, and it would cause issues with our blaze source-file caching.

Reply:
Done.